### PR TITLE
Issue #7983:Resolve Pitest Issues - RegexpCheck (2)

### LINKED ISSUE #7983: Resolve Pitest Issues - RegexpCheck (2)
Child issue of https://github.com/checkstyle/checkstyle/issues/7797 ,

https://github.com/checkstyle/checkstyle/blob/f9131f9536c71d963c83aab1550b7a0371aa9ff5/.ci/pitest.sh#L53

### PR DESCRIPTION
**Pull request by akki111singh**
fixes #7983 

Regression Report: https://akki111singh.github.io/diff/index.html using various configs

### CODE REVIEW

**Comment by strkkk** [comment]
@akki111singh please rebase to latest master to fix CI and resolve merge conflicts

**Comment by akki111singh** [comment]
@strkkk  can you please help me with wercker build failure

**Comment by strkkk** [comment]
I restarted it.

**Comment by akki111singh** [comment]
@strkkk  it failed again

**Comment by akki111singh** [comment]
@strkkk @romani can you please look into the wrecker build failure

**Comment by romani** [comment]
@akki111singh , please rebase to fix CI and resolve conflict

**Comment by romani** [comment]
I relaunched wercker, as it stuck at hbase, looked as unrelated.

File: .ci/pitest.sh [hunk 1]
```diff
@@ -47,9 +47,7 @@ pitest-annotation|pitest-design \
 
 pitest-regexp)
```
Comment (romani) [inline .ci/pitest.sh:48]: please append this label to previous "case".
We do not need this case block any more.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java [hunk 2]
```diff
@@ -627,12 +627,10 @@ else if (!illegalPattern && matchCount == 0) {
...
      * Check if we can stop validation.
      *
-     * @param ignore flag
      * @return true is we can continue
      */
-    private boolean canContinueValidation(boolean ignore) {
-        return errorCount <= errorLimit - 1
-                && (ignore || illegalPattern || checkForDuplicates);
+    private boolean canContinueValidation() {
+        return errorCount <= errorLimit - 1;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java:633]: as we update code, we need to provide regression testing report on all projects.
Please make sure that config that will choose will use updated code.

Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java:633]: Regression report and code logic can be found here
#7983

File: .ci/pitest.sh (hunk 1 above)
Comment (akki111singh) [reply .ci/pitest.sh]: updated

**Comment by romani** [comment]
wercker was failed as:
```
[INFO] --- maven-site-plugin:3.3:site (default-site) @ sample ---
[INFO] configuring report plugin org.apache.maven.plugins:maven-site-plugin:3.3
......
[INFO] Generating "Checkstyle" report    --- maven-checkstyle-plugin:3.1.1
```
strange, restarted.

**Comment by akki111singh** [comment]
@rnveach @romani i have no idea why the wercker build keep on failing everytime

**Comment by rnveach** [comment]
Please try running 'NoExceptiontest - Hbase' locally. CI is reporting error there but log output does not show anything. I will try re-running the job.

**Comment by akki111singh** [comment]
@rnveach How to run NoExceptiontest - Hbase locally?

**Comment by rnveach** [comment]
Rebased to try and solve wercker failure.

**Comment by rnveach** [comment]
Rebase did not help. Command keeps timing out.

https://github.com/checkstyle/checkstyle/blob/c4ddefeda37863ffef37a2069e249accfdf40716/.ci/travis/travis.sh#L389
https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/checks-nonjavadoc-error.xml#L503

I can only assume we are timing out due to a possible infinite loop with changes in regexp.

**Comment by rnveach** [comment]
@akki111singh Do you still wish to look into this?

**Comment by rnveach** [comment]
If we lose contact with original authory, I have been looking into this some more and I think the code needs to be re-written.

So far I came up with the following as the first pass:
````
    private void findMatch() {
        final boolean foundMatch = matcher.find();
        if (foundMatch) {
            final FileText text = getFileContents().getText();
            final LineColumn start = text.lineColumn(matcher.start());
            final int startLine = start.getLine();

            final boolean ignore = isIgnore(startLine, text, start);

            if (!ignore) {
                matchCount++;
                if (illegalPattern || checkForDuplicates
                        && matchCount - 1 > duplicateLimit) {
                    errorCount++;
                    logMessage(startLine);

                    if (errorCount <= errorLimit - 1) {
                        findMatch();
                    }
                }
                else if (checkForDuplicates) { // pitest survives
                    findMatch();
                }
            }
            else {
                findMatch();
            }
        }
        else if (!illegalPattern && matchCount == 0) {
            logMessage(0);
        }
    }
````
It narrows down the number of surviving mutations from 3 to 1. The if conditions is basically inlined to each `findMatch()`.

I would like to see the code refactored further where we have a loop instead of a recursive method call.

**Comment by romani** [comment]
wercker failure related to fix:
```
Command: ./.ci/wercker.sh no-exception-hbase
CS_version: 8.33-SNAPSHOT
Cloning into 'contribution'...
Testing Checkstyle started
   [delete] Deleting directory /pipeline/source/.ci-temp/contribution/checkstyle-tester/src/main/java
Cloning git repository 'Hbase' to repositories/Hbase folder ...
Cloning into 'repositories/Hbase'...
Cloning git repository 'Hbase' - completed

Reseting git sources to commit '1.1.0.1'
Running command: git reset --hard 1.1.0.1
HEAD is now at 4de7d45cb5 Update version in POMs to 1.1.0.1
Hbase is synchronized
     [copy] Copying 3216 files to /pipeline/source/.ci-temp/contribution/checkstyle-tester/src/main/java/Hbase
Running 'mvn clean' on src/main/java ...
Running command: mvn --batch-mode clean
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------< com.github.sevntu.checkstyle:sample >-----------------
[INFO] Building sample 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ sample ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.349 s
[INFO] Finished at: 2020-05-28T16:11:08Z
[INFO] ------------------------------------------------------------------------
Running Checkstyle on src/main/java ... with excludes {}
Running command: mvn -e --batch-mode site -Dcheckstyle.config.location=checks-nonjavadoc-error.xml -Dcheckstyle.excludes= -Dcheckstyle.version=8.33-SNAPSHOT
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------< com.github.sevntu.checkstyle:sample >-----------------
[INFO] Building sample 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-site-plugin:3.3:site (default-site) @ sample ---
[INFO] configuring report plugin org.apache.maven.plugins:maven-site-plugin:3.3
[INFO] configuring report plugin org.apache.maven.plugins:maven-project-info-reports-plugin:2.7
[INFO] configuring report plugin org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1
[INFO] configuring report plugin org.apache.maven.plugins:maven-jxr-plugin:2.5
[WARNING] No project URL defined - decoration links will not be relativized!
[INFO] Rendering site with org.apache.maven.skins:maven-default-skin:jar:1.0 skin.
[INFO] Skipped "Source Xref" report, file "xref/index.html" already exists for the English version.
[INFO] Generating "About" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Plugin Management" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Distribution Management" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Dependency Information" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Source Repository" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Mailing Lists" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Issue Tracking" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Continuous Integration" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Project Plugins" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Project License" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Project Team" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Project Summary" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Dependencies" report    --- maven-project-info-reports-plugin:2.7
[INFO] Generating "Checkstyle" report    --- maven-checkstyle-plugin:3.1.1
```

**Comment by romani** [comment]
@akki111singh , please try to find time to finish this PR.

**Comment by akki111singh** [comment]
@romani  sorry for the dealy! i have been very busy during this work of time.But i will try to find time and look into this.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/pull/8883 is close to completion to make it in right way

**Pull request closed without merging**